ml_genn/model.py

```diff
@@ -290,108 +303,229 @@ def convert_tf_model(tf_model, converter=Simple(),
-                    layer = AvePool2DConv2D(
-                        name=tf_layer.name, filters=tf_layer.filters,
-                        pool_size=pool_layer.pool_size, conv_size=tf_layer.kernel_size,
-                        pool_strides=pool_layer.strides, conv_strides=tf_layer.strides,
-                        pool_padding=pool_layer.padding, conv_padding=tf_layer.padding,
-                        connectivity_type=connectivity_type, 
-                        neurons=converter.create_neurons(tf_layer, pre_compile_output))
+                    tf_in_layers.append(in_layer)
+            tf_in_layers_all[tf_layer] = tf_in_layers
 
-                layer.connect([previous_layer])
-                layer.set_weights(tf_layer.get_weights())
+            # find outbound layers
+            tf_out_layers = [n.outbound_layer for n in tf_layer.outbound_nodes
+                             if n in tf_model_nodes]
+            tf_out_layers_all[tf_layer] = tf_out_layers
 
-                model.layers.append(layer)
-                previous_layer = layer
-                pool_layer = None
 
-            # === AveragePooling2D Layers ===
-            elif isinstance(tf_layer, tf.keras.layers.AveragePooling2D):
-                print('deferring AveragePooling2D layer <{}>'.format(tf_layer.name))
+        # Perform any pre-compilation tasks
+        pre_compile_output = converter.pre_compile(tf_model)
 
-                pool_layer = tf_layer
+        # configure ML GeNN model build process
+        mlg_model_inputs = []
+        mlg_model_outputs = []
+        mlg_layer_lookup = {}
+        new_tf_layers = set()
+        traversed_tf_layers = set()
+
+        if isinstance(tf_model, tf.keras.models.Sequential):
+            tf_in_layers = tf_in_layers_all[tf_model.layers[0]]
+            assert(len(tf_in_layers) == 1)
+            tf_out_layers = [n.outbound_layer for n in tf_in_layers[0].outbound_nodes
+                             if n in tf_model_nodes]
+            tf_out_layers_all[tf_in_layers[0]] = tf_out_layers
+
+        else:
+            tf_in_layers = [tf_model.get_layer(name) for name in tf_model.input_names]
+
+        # === Input Layers ===
+        for tf_layer in tf_in_layers:
+            new_tf_layers.add(tf_layer)
+
+            print('configuring Input layer <{}>'.format(tf_layer.name))
+
+            name = tf_layer.name
+            assert(len(tf_layer.input_shape) == 1)
+            shape = tf_layer.input_shape[0][1:]
+
+            # create layer
+            mlg_layer = InputLayer(name=name, shape=shape,
+                neurons=converter.create_input_neurons(pre_compile_output))
+
+            mlg_layer_lookup[tf_layer] = mlg_layer
+            mlg_model_inputs.append(mlg_layer)
+
+        # while there are still layers to traverse
+        while new_tf_layers:
+            new_tf_layer = new_tf_layers.pop()
+            traversed_tf_layers.add(new_tf_layer)
+
+            # get next TF layer to configure
```

Comment:
Why not follow more of a classic breadth-first-search/traversal:
1. Add the inputs to ``new_tf_layers`` outside the while loop
2. In the while loop, first of all, handle processing of all layer types including input
3. Add outputs to the ``new_tf_layers``

I may be missing something, but it would make the algorithm a little clearer to me

Reply:
I need *all* upstream layers to be created before I create the new one. Breadth first search may not guarantee this in all cases.

Reply:
I'm guessing those cases come from when you have multiple inputs?

Reply:
For instance.